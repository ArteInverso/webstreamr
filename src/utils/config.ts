import { Config } from '../types';

export const getDefaultConfig = (): Config => {
  return { multi: 'on', en: 'on' };
};

export const showExternalUrls = (config: Config): boolean => 'includeExternalUrls' in config;

export const hasMultiEnabled = (config: Config): boolean => 'multi' in config;
