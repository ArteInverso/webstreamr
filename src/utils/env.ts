export const envGet = (name: string): string | undefined => process.env[name];

export const envGetAppId = (): string => process.env['MANIFEST_ID'] || 'webstreamr';

export const envGetAppName = (): string => process.env['MANIFEST_NAME'] || 'WebStreamr';

export const envIsProd = (): boolean => process.env['NODE_ENV'] === 'production';
