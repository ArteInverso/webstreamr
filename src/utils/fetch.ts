import makeFetchHappen from 'make-fetch-happen';
import { FetchInterface, FetchOptions } from 'make-fetch-happen';
import fs from 'node:fs';
import * as os from 'node:os';
import { Response } from 'node-fetch';
import { logInfo } from './log';

// eslint-disable-next-line @typescript-eslint/no-extraneous-class
class MakeFetchHappenSingleton {
  private static instance: FetchInterface;

  public static getInstance(): FetchInterface {
    if (!MakeFetchHappenSingleton.instance) {
      MakeFetchHappenSingleton.instance = makeFetchHappen.defaults({
        cachePath: `${fs.realpathSync(os.tmpdir())}/webstreamr`,
      });
    }
    return MakeFetchHappenSingleton.instance;
  }
}

const cachedFetch = async (uriOrRequest: string | Request, opts?: FetchOptions): Promise<Response> => {
  logInfo(`Fetch ${uriOrRequest}`);

  const fetch = MakeFetchHappenSingleton.getInstance();

  return await fetch(uriOrRequest, opts);
};

export const cachedFetchText = async (uriOrRequest: string | Request, opts?: FetchOptions): Promise<string> => {
  const response = await cachedFetch(uriOrRequest, opts);
  if (!response.ok) {
    throw new Error(`HTTP error: ${response.status} - ${response.statusText}`);
  }

  return await response.text();
};
