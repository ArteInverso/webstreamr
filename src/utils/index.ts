export * from './env';
export * from './Fetcher';
export * from './StreamResolver';
export * from './embed';
export * from './imdb';
export * from './languageFromCountryCode';
export * from './manifest';
export * from './tmdb';
