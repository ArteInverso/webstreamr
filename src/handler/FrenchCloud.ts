import * as cheerio from 'cheerio';
import { Handler } from './types';
import { Fetcher, parseImdbId } from '../utils';
import { ExtractorRegistry } from '../extractor';
import { Context, CountryCode } from '../types';

export class FrenchCloud implements Handler {
  readonly id = 'frenchcloud';

  readonly label = 'FrenchCloud';

  readonly contentTypes = ['movie'];

  readonly countryCodes: CountryCode[] = ['fr'];

  private readonly fetcher: Fetcher;
  private readonly extractorRegistry: ExtractorRegistry;

  constructor(fetcher: Fetcher, extractorRegistry: ExtractorRegistry) {
    this.fetcher = fetcher;
    this.extractorRegistry = extractorRegistry;
  }

  readonly handle = async (ctx: Context, _type: string, id: string) => {
    if (!id.startsWith('tt')) {
      return [];
    }

    const pageUrl = new URL(`https://frenchcloud.cam/movie/${parseImdbId(id).id}`);
    const html = await this.fetcher.text(ctx, pageUrl);

    const $ = cheerio.load(html);

    return Promise.all(
      $('[data-link!=""]')
        .map((_i, el) => new URL(($(el).attr('data-link') as string).replace(/^(https:)?\/\//, 'https://')))
        .toArray()
        .filter(url => !url.host.match(/frenchcloud/))
        .map(url => this.extractorRegistry.handle({ ...ctx, referer: pageUrl }, url, { countryCode: 'fr' })),
    );
  };
}
