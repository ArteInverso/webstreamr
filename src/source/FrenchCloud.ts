import { ContentType } from 'stremio-addon-sdk';
import * as cheerio from 'cheerio';
import { Source, SourceResult } from './types';
import { Fetcher, getImdbId, Id } from '../utils';
import { Context, CountryCode } from '../types';

export class FrenchCloud implements Source {
  public readonly id = 'frenchcloud';

  public readonly label = 'FrenchCloud';

  public readonly contentTypes: ContentType[] = ['movie'];

  public readonly countryCodes: CountryCode[] = [CountryCode.fr];

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  public readonly handle = async (ctx: Context, _type: string, id: Id): Promise<SourceResult[]> => {
    const imdbId = await getImdbId(ctx, this.fetcher, id);

    const pageUrl = new URL(`https://frenchcloud.cam/movie/${imdbId.id}`);
    const html = await this.fetcher.text(ctx, pageUrl);

    const $ = cheerio.load(html);

    return Promise.all(
      $('[data-link!=""]')
        .map((_i, el) => new URL(($(el).attr('data-link') as string).replace(/^(https:)?\/\//, 'https://')))
        .toArray()
        .filter(url => !url.host.match(/frenchcloud/))
        .map(url => ({ countryCode: CountryCode.fr, referer: pageUrl, url })),
    );
  };
}
