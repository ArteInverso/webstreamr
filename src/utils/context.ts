import { Request } from 'express';
import { Context } from '../types';
import { getDefaultConfig } from './config';

export const contextFromRequest = (req: Request): Context => {
  return {
    hostUrl: new URL(`${req.protocol}://${req.host}`),
    id: req.headers['X-Request-ID'] as string,
    ...(req.ip && { ip: req.ip }),
    config: req.params['config'] ? JSON.parse(req.params['config']) : getDefaultConfig(),
  };
};
