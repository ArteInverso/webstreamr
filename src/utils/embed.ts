import { unpack } from 'unpacker';
import { cachedFetchText } from './fetch';
import { scanFromResolution } from './resolution';

const LINK_REGEXPS = [
  /sources:\[{file:"(.*?)"/, // dropload, supervideo
];

interface ParsePackedEmbedResult {
  url: string;
  resolution: string | undefined;
  size: string | undefined;
}

const findResolution = (html: string): string | undefined => {
  const match = html.match(/(\d{3,}x\d{3,}),/);

  return match && match[1] ? scanFromResolution(match[1]) : undefined;
};

const findSize = (html: string): string | undefined => {
  const sizeMatch = html.match(/([\d.]+) ?([GM]B)/);

  return sizeMatch && sizeMatch[1] && sizeMatch[2] ? `${sizeMatch[1]} ${sizeMatch[2]}` : undefined;
};

export const parsePackedEmbed = async (url: string): Promise<ParsePackedEmbedResult> => {
  const html = await cachedFetchText(url);

  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found on ${url}`);
  }

  const unpacked = unpack(evalMatch[0]);

  for (const linkRegexp of LINK_REGEXPS) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return {
        url: 'https://' + linkMatch[1].replace(/^(https:)?\/\//, ''),
        resolution: findResolution(html),
        size: findSize(html),
      };
    }
  }

  throw new Error(`Could not find a stream link on ${url}`);
};
