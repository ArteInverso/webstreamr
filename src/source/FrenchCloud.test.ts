import { createTestContext } from '../test';
import { FetcherMock, ImdbId } from '../utils';
import { FrenchCloud } from './FrenchCloud';

const ctx = createTestContext({ fr: 'on' });

describe('FrenchCloud', () => {
  let source: FrenchCloud;

  beforeEach(() => {
    source = new FrenchCloud(new FetcherMock(`${__dirname}/__fixtures__/FrenchCloud`));
  });

  test('handles non-existent movies gracefully', async () => {
    const streams = await source.handle(ctx, 'movie', new ImdbId('tt12345678', undefined, undefined));
    expect(streams).toHaveLength(0);
  });

  test('handle imdb the devil\'s bath', async () => {
    const streams = await source.handle(ctx, 'movie', new ImdbId('tt29141112', undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
