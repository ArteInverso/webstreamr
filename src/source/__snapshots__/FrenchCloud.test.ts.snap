// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FrenchCloud handle imdb the devil's bath 1`] = `
[
  {
    "countryCode": "fr",
    "referer": "https://frenchcloud.cam/movie/tt29141112",
    "url": "https://supervideo.cc/e/ttnqt211ebtg",
  },
  {
    "countryCode": "fr",
    "referer": "https://frenchcloud.cam/movie/tt29141112",
    "url": "https://dropload.io/embed-w4yit5wiia9y.html",
  },
  {
    "countryCode": "fr",
    "referer": "https://frenchcloud.cam/movie/tt29141112",
    "url": "https://mixdrop.ag/e/l7v73zqrfdj19z",
  },
  {
    "countryCode": "fr",
    "referer": "https://frenchcloud.cam/movie/tt29141112",
    "url": "https://streamtape.com/e/gjA1OQ4klyHxgJ",
  },
  {
    "countryCode": "fr",
    "referer": "https://frenchcloud.cam/movie/tt29141112",
    "url": "https://dood.to/e/osbcylatzy6m",
  },
]
`;
